File: reader/rust/src/parser.rs
use crate::reader::{MetaType, Reader};
use serde::Serialize;
use tucana::shared::{DefinitionDataType, FlowType, RuntimeFunctionDefinition};

#[derive(Serialize, Clone, Debug)]
pub struct DefinitionError {
    pub definition: String,
    pub definition_type: MetaType,
    pub error: String,
}

#[derive(Debug)]
pub struct Parser {
    pub features: Vec<Feature>,
}

#[derive(Serialize, Clone, Debug)]
pub struct Feature {
    pub name: String,
    pub data_types: Vec<DefinitionDataType>,
    pub flow_types: Vec<FlowType>,
    pub runtime_functions: Vec<RuntimeFunctionDefinition>,
    pub errors: Vec<DefinitionError>,
}

impl Feature {
    fn new(name: String) -> Self {
        Feature {
            name,
            data_types: Vec::new(),
            flow_types: Vec::new(),
            runtime_functions: Vec::new(),
            errors: Vec::new(),
        }
    }
}

impl Parser {

    pub fn from_path(path: &str) -> Option<Self> {
        let reader = match Reader::from_path(path) {
            Some(reader) => reader,
            None => return None,
        };

        Some(Self::from_reader(reader))
    }

    pub fn from_reader(reader: Reader) -> Self {
        let mut features: Vec<Feature> = vec![];

        for meta in &reader.meta {
            let feature = features.iter_mut().find(|f| f.name == meta.name);

            if let Some(existing) = feature {
                Parser::append_meta(existing, meta);
            } else {
                let mut new_feature = Feature::new(meta.name.clone());
                Parser::append_meta(&mut new_feature, meta);
                features.push(new_feature);
            }
        }

        Parser { features }
    }

    fn append_meta(feature: &mut Feature, meta: &crate::reader::Meta) {
        for definition in &meta.data {
            match meta.r#type {
                MetaType::DataType => {
                    match serde_json::from_str::<DefinitionDataType>(definition) {
                        Ok(data_type) => feature.data_types.push(data_type),
                        Err(err) => feature.errors.push(DefinitionError {
                            definition: definition.to_string(),
                            definition_type: MetaType::DataType,
                            error: err.to_string()
                        })
                    }
                }
                MetaType::FlowType => match serde_json::from_str::<FlowType>(definition) {
                    Ok(flow_type) => feature.flow_types.push(flow_type),
                    Err(err) => feature.errors.push(DefinitionError {
                        definition: definition.to_string(),
                        definition_type: MetaType::FlowType,
                        error: err.to_string()
                    })
                },
                MetaType::RuntimeFunction => {
                    match serde_json::from_str::<RuntimeFunctionDefinition>(definition) {
                        Ok(func) => feature.runtime_functions.push(func),
                        Err(err) => feature.errors.push(DefinitionError {
                            definition: definition.to_string(),
                            definition_type: MetaType::RuntimeFunction,
                            error: err.to_string()
                        })
                    }
                }
            }
        }
    }
}

File: reader/ts/src/parser.ts
import {DefinitionDataType as TucanaDataType} from "@code0-tech/tucana/pb/shared.data_type_pb.js";
import {Feature} from "./types.js";
import {readdirSync, readFileSync} from "node:fs";
import {FlowType as TucanaFlowType} from "@code0-tech/tucana/pb/shared.flow_definition_pb.js";
import {RuntimeFunctionDefinition as TucanaFunction} from "@code0-tech/tucana/pb/shared.runtime_function_pb.js";
import path from "node:path";
import {mapFlowType} from "./mapper/flowTypeMapper.js";
import {mapFunction} from "./mapper/functionMapper.js";
import {DataType} from "@code0-tech/sagittarius-graphql-types";
import {DefinitionDataType} from "@code0-tech/tucana/pb/shared.data_type_pb.js";
import {getDataType} from "./mapper/dataTypeMapper.js";

export interface ConstructedDataTypes {
    scannedTucanaTypes: DefinitionDataType[]
    constructedDataTypes: DataType[]
    id: number
}

export function getID(constructedDataTypes: ConstructedDataTypes) {
    const last = constructedDataTypes.id
    constructedDataTypes.id += 1
    return last
}

export const Definition = (rootPath: string): Feature[] => {
    const dataTypes: {feature: string, type: TucanaDataType}[] = []
    const runtimeFunctions: {feature: string, func: TucanaFunction}[] = [];
    const flowTypes: {feature: string, flow: TucanaFlowType}[] = [];

    readdirSync(rootPath, { withFileTypes: true }).forEach(file => {
        const featureName = file.name.split("_")[0]
        if (featureName == null) {
            throw new Error("Feature name is null")
        }

        const filePath = path.join(file.parentPath, file.name)

        const content = readFileSync(filePath);
        if (file.name.includes("data_type")) {
            const decoded = TucanaDataType.fromBinary(content);
            dataTypes.push(
                {
                    feature: featureName,
                    type: decoded,
                }
            )
        }

        if (file.name.includes("function")) {
            const decoded = TucanaFunction.fromBinary(content);
            runtimeFunctions.push(
                {
                    feature: featureName,
                    func: decoded,
                }
            )
        }

        if (file.name.includes("flow_type")) {
            const decoded = TucanaFlowType.fromBinary(content);
            flowTypes.push(
                {
                    feature: featureName,
                    flow: decoded,
                }
            )
        }
    })

    const features: Feature[] = []
    const constructed: ConstructedDataTypes = {
        scannedTucanaTypes: dataTypes.map(f => f.type),
        constructedDataTypes: [],
        id: 0
    }

    function getFeature(name:string): Feature {
        const feature = features.find((f) => f.name === name);
        if (feature != undefined) {
            return feature;
        }

        const newFeature = {
            name: name,
            dataTypes: [],
            flowTypes: [],
            runtimeFunctions: [],
        };

        features.push(newFeature);
        return newFeature;
    }

    dataTypes.map(f => {
        return {
            name: f.feature,
            type: getDataType(f.type.identifier, constructed)
        }
    }).forEach(dt => {
        if (dt.type != null) {
            const feature = getFeature(dt.name)
            feature.dataTypes.push(dt.type)
        }
    })

    runtimeFunctions.map(f => {
        return {
            name: f.feature,
            type: mapFunction(f.func, constructed)
        }
    }).forEach(dt => {
        if (dt.type != null) {
            const feature = getFeature(dt.name)
            feature.runtimeFunctions.push(dt.type)
        }
    })

    flowTypes.map(f => {
        return {
            name: f.feature,
            type: mapFlowType(f.flow, constructed)
        }
    }).forEach(dt => {
        if (dt.type != null) {
            const feature = getFeature(dt.name)
            feature.flowTypes.push(dt.type)
        }
    })

    return features;
}
